java-checks/src/main/java/org/sonar/java/checks/ShiftOnIntOrLongCheck.java

```diff
@@ -0,0 +1,114 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.check.BelongsToProfile;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.model.AbstractTypedTree;
+import org.sonar.java.resolve.Type;
+import org.sonar.plugins.java.api.tree.ArrayAccessExpressionTree;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+import org.sonar.plugins.java.api.tree.UnaryExpressionTree;
+
+import java.text.MessageFormat;
+import java.util.List;
+
+@Rule(
+  key = "S2183",
+  priority = Priority.CRITICAL,
+  tags = {"bug"})
+@BelongsToProfile(title = "Sonar way", priority = Priority.CRITICAL)
+public class ShiftOnIntOrLongCheck extends SubscriptionBaseVisitor {
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Kind.LEFT_SHIFT, Kind.LEFT_SHIFT_ASSIGNMENT, Kind.RIGHT_SHIFT, Kind.RIGHT_SHIFT_ASSIGNMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ExpressionTree target = null;
+    ExpressionTree shift = null;
+
+    if (tree.is(Kind.LEFT_SHIFT, Kind.RIGHT_SHIFT)) {
+      BinaryExpressionTree binaryExpressionTree = (BinaryExpressionTree) tree;
+      target = binaryExpressionTree.leftOperand();
+      shift = binaryExpressionTree.rightOperand();
+    } else {
+      AssignmentExpressionTree assignmentExpressionTree = (AssignmentExpressionTree) tree;
+      target = assignmentExpressionTree.variable();
+      shift = assignmentExpressionTree.expression();
+    }
+
+    String identifier = getIdentifierName(target);
+    Type expectedType = ((AbstractTypedTree) target).getSymbolType();
+    boolean expectInt = expectedType.is("int") || expectedType.is("java.lang.Integer");
+    boolean shiftIsMinus = shift.is(Kind.UNARY_MINUS);
+
+    if (shift.is(Kind.UNARY_MINUS, Kind.UNARY_PLUS)) {
+      shift = ((UnaryExpressionTree) shift).expression();
+    }
+
+    if (shift.is(Kind.INT_LITERAL, Kind.LONG_LITERAL)) {
+      String value = ((LiteralTree) shift).value();
+      long numberBits = (shiftIsMinus ? -1 : 1) * Long.parseLong(value);
+      long reducedNumberBits = numberBits % (expectInt ? 32 : 64);
+      insertIssue(tree, numberBits, reducedNumberBits, expectInt, identifier);
+    }
+  }
+
+  private void insertIssue(Tree tree, long numberBits, long reducedNumberBits, boolean expectInt, String identifier) {
```

Comment:
I'm not sure of it, as the method does not necessarily add issue to the tree. However, I can change it to return a message (potentially null when no message should be added) and test the return value before adding the issue.